source/solvers/vof.cc

```diff
@@ -732,7 +791,268 @@ VOF<dim>::solve_linear_system(const bool initial_step,
+                  << linear_solver_tolerance << std::endl;
+    }
+
+  TrilinosWrappers::MPI::Vector completely_distributed_phase_fraction_solution(
+    this->locally_owned_dofs, triangulation->get_communicator());
+
+
+  SolverControl solver_control(
+    this->simulation_parameters.linear_solver.max_iterations,
+    linear_solver_tolerance,
+    true,
+    true);
+
+  TrilinosWrappers::SolverCG solver(solver_control);
+
+  //**********************************************
+  // Trillinos Wrapper ILU Preconditioner
+  //*********************************************
+  const double ilu_fill =
+    this->simulation_parameters.linear_solver.ilu_precond_fill;
+  const double ilu_atol =
+    this->simulation_parameters.linear_solver.ilu_precond_atol;
+  const double ilu_rtol =
+    this->simulation_parameters.linear_solver.ilu_precond_rtol;
+
+  TrilinosWrappers::PreconditionILU::AdditionalData preconditionerOptions(
+    ilu_fill, ilu_atol, ilu_rtol, 0);
+
+  ilu_preconditioner = std::make_shared<TrilinosWrappers::PreconditionILU>();
+
+  ilu_preconditioner->initialize(system_matrix_phase_fraction,
+                                 preconditionerOptions);
+
+  solver.solve(system_matrix_phase_fraction,
+               completely_distributed_phase_fraction_solution,
+               system_rhs_phase_fraction,
+               *ilu_preconditioner);
+
+  if (this->simulation_parameters.interface_sharpening.verbosity !=
+      Parameters::Verbosity::quiet)
+    {
+      this->pcout << "  -Iterative solver took : " << solver_control.last_step()
+                  << " steps " << std::endl;
+    }
+
+  nonzero_constraints.distribute(
+    completely_distributed_phase_fraction_solution);
+  solution = completely_distributed_phase_fraction_solution;
+}
+
+template <int dim>
+void
+VolumeOfFluid<dim>::assemble_mass_matrix_diagonal(
+  TrilinosWrappers::SparseMatrix &mass_matrix)
+{
+  QGauss<dim> quadrature_formula(this->cell_quadrature->size());
+
+  FEValues<dim> fe_values(*fe,
+                          quadrature_formula,
+                          update_values | update_JxW_values);
+
+
+  const unsigned int dofs_per_cell = this->fe->dofs_per_cell;
+  const unsigned int n_qpoints     = quadrature_formula.size();
+  FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+  std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+  for (const auto &cell : dof_handler.active_cell_iterators())
+    {
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit(cell);
+          cell_matrix = 0;
+          for (unsigned int q = 0; q < n_qpoints; ++q)
+            for (unsigned int i = 0; i < dofs_per_cell; ++i)
+              cell_matrix(i, i) +=
+                (fe_values.shape_value(i, q) * fe_values.shape_value(i, q) *
+                 fe_values.JxW(q));
+          cell->get_dof_indices(local_dof_indices);
+          nonzero_constraints.distribute_local_to_global(cell_matrix,
```

Comment:
Add comment to explain what is being assembled.

Reply:
This function is explained in detail in step-41 (with the same name). I've copied and added some explanations to make it more clear.